File: path_planner.py
import numpy as np
import matplotlib.pyplot as plt
from simulator import *
from path_controller import *
import pdb

def plan(mission, x):
	n=len(mission)
	if n==3: path=brute_force(mission, x)
	if n==4: path=DP_dyn_sort(mission, x)
	if n==5: path=hybrid(mission, x)
	
	return path
	


#----------------DP
# input: mission as a list of arrays
#        x - state of plane (array)
# output: mission_sorted
# variables: W matrix contrinaing all the wieght for the DP alorithms


def DP_sort(mission, x):

	n=len(mission)
	W=np.zeros((n,n+1))
	path=[0] #structure of path i is [mission points, cost]
	cost=0.
	for i in range(n):
		W[i,0]=np.linalg.norm(x-mission[i])
		for j in range(n):
			W[i,j+1]=np.linalg.norm(mission[i]-mission[j])

	for i in range(n):
		i+=1 # cause 0 is starting point
		opt=np.ones(i)
		opt=opt*cost
		for j in range(i):
			opt[j]+=W[i-1,path[j]]

		cost=min(opt)
		path.insert(np.argmin(opt)+1, i)

	return path
	
#return: path=list[(x,y,z) at t=computation_starts, (x,y,z) of mission point in optimized order]
#computational cost: n! * (simulation+cost_function) + n (find min)

def DP_dyn_sort(mission, x):

	n=len(mission)
	path=[x[:2]]
	cost=0.

	for i in range(n):
		i+=1 # cause 0 is starting point
		opt=np.ones(i)
		opt=opt*cost
		
		for j in range(i):
			path_opt=path.insert(j+1, mission[i])
			controller=PathController(path_opt)
			stats=rollout(controller.policy, x, 100, .1)
			pos=np.vstack(stats["x"])
			x=pos[:,0]
			y=pos[:,1]
			z=pos[:,2]
			w=0.
			for i in range(len(x)-1): 
				w+=np.sqrt((x[i]-x[i+1])**2.+(y[i]-y[i+1])**2.+(z[i]-z[i+1])**2.)
			opt[j]+=w

		cost=min(opt)
		path.insert(np.argmin(opt)+1, mission[i])

	return path
	
	
def brute_force(mission, x):

	if len(mission)!=3: break

	cost=np.ones(6)
	order=np.array([1,2,3;1,3,2;2,3,1;2,1,3;3,1,2;3,2,1])
	for k in range(6):
		path_opt=[mission[j-1] for j in order[k, :]]
		path_opt=path_opt.insert(0,x[:2])
		
		controller=PathController(path_opt)
		stats=rollout(controller.policy, x, 100, .1)
		pos=np.vstack(stats["x"])
		x=pos[:,0]
		y=pos[:,1]
		z=pos[:,2]
		w=0.
		for i in range(len(x)-1): 
			w+=np.sqrt((x[i]-x[i+1])**2.+(y[i]-y[i+1])**2.+(z[i]-z[i+1])**2.)
		cost[k]=w
		
	idx=min(cost)
	path= [mission[j-1] for j in order(idx, :)]
	
	return path.insert(0, x[:2])
	
def hybrid(mission, x):

	idx=DP_sort(mission, x)
	cost=[]
	IJK=np.array([2,2,3,3,4,4;3,4,2,4,2,3;4,3,4,2,3,2])
	for i,j,k in IJK[0,:], IJK[1,:], IJK[2,:]:
		path_opt=x[:2]+mission[idx[1]]+mission[i]+mission[j]+mission[k]+mission[-1] #init, closest point, ...,farest point
		
		controller=PathController(path_opt)
		stats=rollout(controller.policy, x, 100, .1)
		pos=np.vstack(stats["x"])
		x=pos[:,0]
		y=pos[:,1]
		z=pos[:,2]
		w=0.
		for i in range(len(x)-1): 
			w+=np.sqrt((x[i]-x[i+1])**2.+(y[i]-y[i+1])**2.+(z[i]-z[i+1])**2.)
		cost.append(w)
	
	i=min(cost)
	path=x[:2]+mission[idx[1]]+mission[[0,i]]+mission[IJK[1,i]]+mission[IJK[2,i]]+mission[-1]
	
	return path
	

#----------------------------------
# input: mission as it was (only pts ahead, not the passed ones!), x is current state, new point new point
# output: path -list of new points

#----------------------------------

def DP_add_pt(mission, x, new_pts):
	
	m=len(mission)
	n=len(new_pts)
	W=np.zeros((m,n+1+m))
	x=x[:2]
	path=mission #structure of path i is [mission points, cost]
	cost=0.

	for i in range(n):
		W[i,0]=np.linalg.norm(x-new_pts[i])
		for j in range(m):
			W[i,j+1]=np.linalg.norm(new_pts[i]-mission[j])
		for k in range(n):
			W[i,m+1+k]=np.linalg.norm(new_pts[i]-new_pts[k])

	for i in range(n):
		i+=m+1
		opt=np.ones(i)
		opt=opt*cost
		for j in range(i):
			opt[j]+=W[i-(n+1),j]

		cost=min(opt)
		path.insert(np.argmin(opt)+1, new_pts[i-(n+1)])

	return path

	

File: path_planner.cpp
#include <iostream.h>
#include <stdlib.h>
#include <math.h>

using namespace std

//based on distance + penalty for high angles
double weight(double x1[], double x2[], double v, double x_prev[])
{
	//distance eval
	double l1=x2[0]-x1[0];
	double l2=x2[1]-x1[1];
	double v1[]={l1, l2};
	w_d=sqrt(pow(l1,2)+pow(l2,2));
	
	const double pi= 3.141592653589793238462643383279502884;
	
	//angle
	double v2[]={x_prev[0]-x1[0],x_prev[1]-x1[1]};
	double theta=(180.0/pi)*acos((v1[0]*v2[0]+v1[1]*v2[1]) /(pow(v1[0],2)+pow(v1[1],2))/ (pow(v2[0],2)+pow(v2[1],2)));
	double w_angel=0.0;
	
	//penalty 
	const double p=20.0;
	
	if (w_angel<120.0){
		w_angle+=p;
		};
	if (w_angel<60.0){
		w_angle+=p;
		};
	if (w_angel<30.0){
		w_angle+=pow(10,10);
		};
	
	w=w_d+w_angle
	
	if (x_prev.size()==1) {
		return w_d
		}
	else
		return w;
	}

//find lowest cost among options
int get_lowest(double vec[])
{
	int idx=100;
	double min=1000000;
	for (int i=0; i<vec.size(); i++){
		if (vec[i]<min)
			min=vec[i];
			idx=i;
		}
	return idx;
	}

//inster point pt at position idx in path
double insert(double path[], double pt[], int idx)
{
	const int pl=path.size();
	double new_path[pl+1];
	for (int k=0; k<idx; k++){
		new_path[k]=path[k];
		}
	new_path[idx]=pt;
	for (int k=idx+1; k<pl; k++){
		new_path[k]=path[k-1];
		}
	return new_path;
	}

//input: mission, pos
//output: optimized path 
double DP(double mission[], double pos[])
{
	const int n=mission.size();
	double cost=0;
	double path[]={pos};
	
	//cruise velocity
	const double v=15.5;
	
	for (int k=0; k<n; k++)
	{
		double pt[]=mission[k];
		double options[k+1];
		for (int i=0; i<k+1; i++){
			double path_opt[]=insert(path, mission[k], i+1);
			options[i]=weight(path_opt[0], path_opt[1], v, 1.0);
			for (int j=1; j<k+2; j++){
				options[i]+=weight(path_opt[j], path_opt[j+1], v, path_opt[j-1])
				}
		}
		idx=get_lowest(options)
		path=insert(path, mission[k], idx+1)
	}
	return path;
}
